env-var-docs/parser-package/src/env_var_docs/parser.py

```diff
@@ -0,0 +1,549 @@
+        """Callback for json.load object_pairs_hook parameter."""
+        out = {}
+        for k, v in fields:
+            if k in out:
+                raise ValueError(f"found duplicate key '{k}'")
+            out[k] = v
+        return out
+
+    json_root = "file"
+    try:
+        docs = json.load(file, object_pairs_hook=raise_on_duplicate_keys)
+    except (json.JSONDecodeError, ValueError) as e:
+        return [
+            NodeParseError(
+                json_root, [ParseError(json_root, f"file is not valid: {e}")],
+                [])
+        ]
+    if not isinstance(docs, dict):
+        return [
+            NodeParseError(
+                json_root,
+                [ParseError(json_root, f"file is not a single JSON object")],
+                [])
+        ]
+
+    nodes_to_visit: list[Node] = []
+    parsing_errors: list[NodeParseError] = []
+
+    # The file is an implicit root node. Recursively descend into each top-level field.
+    for key, value in docs.items():
+        _recursively_parse(
+            1, json_root, key, value, nodes_to_visit, parsing_errors)
+
+    if len(parsing_errors) != 0:
+        return parsing_errors
+
+    for node in nodes_to_visit:
+        visit_fn(node)
+
+    return []
+
+
+def _recursively_parse(
+        level: int, parent_path: str, key: str, value: UnparsedJSON,
+        nodes_to_visit: list[Node], parsing_errors: list[NodeParseError]):
+    """Recursively visits a node and its children, in preorder traversal order (visit root then children)."""
+    path = _path(parent_path, key)
+
+    # First, try to parse details as a Group.
+    group, group_errors = _parse_group(path, value)
+    if len(group_errors) == 0:
+        assert group is not None  # appease mypy
+        nodes_to_visit.append(
+            Node(level=level, json_path=parent_path, name=key, details=group))
+
+        # Recurse into group members.
+        for sub_key, sub_value in group.members.items():
+            _recursively_parse(
+                level + 1, path, sub_key, sub_value, nodes_to_visit,
+                parsing_errors)
+        return
+
+    # details was not successfully parsed as a Group. Next, try to parse it as a Variable.
+    var, var_errors = _parse_variable(path, value)
+    if len(var_errors) == 0:
+        assert var is not None  # appease mypy
+        nodes_to_visit.append(
+            Node(level=level, json_path=parent_path, name=key, details=var))
+        return
+
+    # details is neither a group nor a variable.
+    parsing_errors.append(NodeParseError(path, group_errors, var_errors))
+
+
+def _parse_group(parent_path: str,
+                 obj: UnparsedJSON) -> tuple[Group | None, ParseErrors]:
+    """Parses a Group from obj. If ParseErrors are encountered, None is returned."""
+
+    errors = []
+    errors.extend(
+        _ensure_no_extra_fields(
+            parent_path, obj, ["group_description", "members"]))
+
+    group_description, errs = _parse_field(
```

Comment:
Because the schema of the env-var-docs.json file allows for Groups and Variables to be in the same place (top-level fields in the file as well as fields in a Group's members), there is no getting around attempting to parse as both a Variable and a Group.  _recursively_parse handles this by bubbling up ParseErrors only when both _parse_group and _parse_errors return errors